corehq/apps/user_importer/helpers.py

```diff
@@ -252,6 +252,26 @@ def update_locations(self, location_codes, domain_info):
             else:
                 self.logger.add_info(UserChangeMessage.primary_location_removed())
 
+    def update_user_groups(self, domain_info, group_names):
+        old_group_ids = set()
+        for group in domain_info.group_memoizer.by_user_id(self.user.user_id):
+            old_group_ids.add(group.get_id)
+            if group.name not in group_names:
+                group.remove_user(self.user)
+                domain_info.group_memoizer.updated_groups.add(group.get_id)
+
+        new_groups = {}
+        for group_name in group_names:
+            group = domain_info.group_memoizer.by_name(group_name)
+            group.add_user(self.user, save=False)
```

Comment:
can we make `group.add_user` return `True` if the group changed? Then this could become:
```python
if group.add_user(self.user, save=False):
    domain_info.group_memoizer.group_updated(group)
```

Reply:
The idea is to have `group.add_user` tell us if the group changed instead of checking against `old_group_ids`? That makes sense to me. What about `group_memoizer.group_updated(group)`? It's not currently a method on the memoizer. Did you have something in mind for it, or does `group_memoizer.updated_groups.add(group.get_id)` still work?

Reply:
it would still work to do that but you're unnecessarily exposing the `updated_groups` property of the memoizer. Making it a method [encapsulates](https://gamedevelopment.tutsplus.com/tutorials/quick-tip-the-oop-principle-of-encapsulation--gamedev-2187) the functionality in the class.

Reply:
Got it, the memoizer and the importer are now more loosely coupled.